### PR TITLE
[SYCLomatic] Refine the migration of CUDART_VERSION

### PR DESCRIPTION
**Pull request by zhiweij1**
Refine the migration of CUDART_VERSION when use in if/elif directives. For example, the expression like “#if CUDART_VERSION > 12000” is migrate to “#if SYCL_LANGUAGE_VERSION > 202000”.

The migration keeps the update footprint as small as possible by only updating nearest ancestor Boolean operation “node” (E.g.,
'!', '?', '||' or '&&') of the macro CUDART_VERSION. 

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/test/dpct/predefined_macro_replacement.cu [hunk 1]
```diff
@@ -173,3 +173,89 @@ int foo2(){
     pl = CUDART_VERSION - major * 1000 - minor * 10;
   }
 }
+
+//CHECK: void foo3() {
+//CHECK-NEXT: #if (SYCL_LANGUAGE_VERSION >= 202001)
```
Comment (zhimingwang36) [inline clang/test/dpct/predefined_macro_replacement.cu]: ### 
```suggestion
//CHECK-NEXT: #if  SYCL_LANGUAGE_VERSION
```

File: clang/test/dpct/predefined_macro_replacement.cu [hunk 2]
```diff
@@ -173,3 +173,89 @@ int foo2(){
...
+  int2 a1;
+#endif
+#if defined( CUDART_VERSION)
+  int2 a2;
+#endif
+#define AAA 1
+#if (CUDART_VERSION >= 4000) && AAA
+  int2 a3;
+#endif
+#if !CUDART_VERSION && AAA
```
Comment (zhimingwang36) [inline clang/test/dpct/predefined_macro_replacement.cu]:
```suggestion
#if ! (CUDART_VERSION>=4000) && AAA
```

File: clang/test/dpct/predefined_macro_replacement.cu [hunk 3]
```diff
@@ -173,3 +173,89 @@ int foo2(){
...
+  int2 a4;
+#endif
+#if (CUDART_VERSION ? 1 : 0) && AAA
```
Comment (zhimingwang36) [inline clang/test/dpct/predefined_macro_replacement.cu]:
```suggestion
#if (CUDART_VERSION > 4000 ? 1 : 0) && AAA
```

File: clang/test/dpct/predefined_macro_replacement.cu [hunk 4]
```diff
@@ -173,3 +173,89 @@ int foo2(){
...
+
+//CHECK: void foo3() {
+//CHECK-NEXT: #if (SYCL_LANGUAGE_VERSION >= 202001)
+//CHECK-NEXT:   sycl::int2 a1;
+//CHECK-NEXT: #endif
+//CHECK-NEXT: #if (SYCL_LANGUAGE_VERSION >= 202001)
+//CHECK-NEXT:   sycl::int2 a2;
+//CHECK-NEXT: #endif
+//CHECK-NEXT: #define AAA 1
+//CHECK-NEXT: #if (SYCL_LANGUAGE_VERSION >= 202001) && AAA
```
Comment (zhimingwang36) [inline clang/test/dpct/predefined_macro_replacement.cu]: will it have better compatibility among different SYCL compiler if using 202000 ?

File: clang/test/dpct/predefined_macro_replacement.cu (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/predefined_macro_replacement.cu]: done

File: clang/test/dpct/predefined_macro_replacement.cu (hunk 3 above)
Comment (zhiweij1) [reply clang/test/dpct/predefined_macro_replacement.cu]: done

File: clang/test/dpct/predefined_macro_replacement.cu (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/predefined_macro_replacement.cu]: done

File: clang/test/dpct/predefined_macro_replacement.cu (hunk 1 above)
Comment (zhiweij1) [reply clang/test/dpct/predefined_macro_replacement.cu]: done

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/include/clang/Basic/Cuda.h [hunk 5]
```diff
@@ -9,6 +9,9 @@
 #ifndef LLVM_CLANG_BASIC_CUDA_H
 #define LLVM_CLANG_BASIC_CUDA_H
 
+#ifdef SYCLomatic_CUSTOMIZATION
```
Comment (zhimingwang36) [inline clang/include/clang/Basic/Cuda.h]: it is not necessary.

File: clang/lib/Lex/PPDirectives.cpp [hunk 6]
```diff
@@ -717,7 +717,13 @@ void Preprocessor::SkipExcludedConditionalBlock(SourceLocation HashTokenLoc,
           assert(CurPPLexer->LexingRawMode && "We have to be skipping here!");
           CurPPLexer->LexingRawMode = false;
           IdentifierInfo *IfNDefMacro = nullptr;
+#ifdef SYCLomatic_CUSTOMIZATION
```
Comment (zhimingwang36) [inline clang/lib/Lex/PPDirectives.cpp:720]: Please summary the change to lex in the commit message.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 7]
```diff
@@ -52,6 +52,13 @@ using namespace clang::ast_matchers;
 using namespace clang::dpct;
 using namespace clang::tooling;
 
+namespace clang {
+extern bool ProcessingCudaRTVersionMacro;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp:57]: add comment on usage of this flag.

File: clang/include/clang/Basic/Cuda.h (hunk 5 above)
Comment (zhiweij1) [reply clang/include/clang/Basic/Cuda.h]: fixed

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 7 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp:57]: done

File: clang/lib/Lex/PPDirectives.cpp (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/Lex/PPDirectives.cpp:720]: done

File: clang/lib/Lex/Preprocessor.cpp [hunk 8]
```diff
@@ -825,6 +825,12 @@ bool Preprocessor::HandleIdentifier(Token &Identifier) {
 
   // If this is a macro to be expanded, do it.
   if (MacroDefinition MD = getMacroDefinition(&II)) {
+#ifdef SYCLomatic_CUSTOMIZATION
+    if (IsEvaluatingExpression && II.getName().str() == "CUDART_VERSION" &&
+        IsInAnalysisScopeFunc(Identifier.getLocation())) {
```
Comment (zhiweij1) [inline clang/lib/Lex/Preprocessor.cpp:830]:
```suggestion
        IsInRoot(Identifier.getLocation())) {
```

File: clang/lib/Lex/PPExpressions.cpp [hunk 9]
```diff
@@ -226,6 +246,16 @@ static bool EvaluateDefined(PPValue &Result, Token &PeekTok, DefinedTracker &DT,
       PP.Diag(beginLoc, diag::warn_defined_in_object_type_macro);
   }
 
+#ifdef SYCLomatic_CUSTOMIZATION
```
Comment (zhiweij1) [inline clang/lib/Lex/PPExpressions.cpp]: define expr can use callback

File: clang/lib/DPCT/DPCT.cpp [hunk 10]
```diff
@@ -219,6 +220,10 @@ std::string getCudaInstallPath(int argc, const char **argv) {
       Driver, llvm::Triple(Driver.getTargetTriple()), ParsedArgs);
 
   std::string Path = CudaIncludeDetector.getInstallPath().str();
+  uint64_t CudaRTVersionValueUint64 =
+      clang::CudaVersionToValue(CudaIncludeDetector.version());
```
Comment (zhiweij1) [inline clang/lib/DPCT/DPCT.cpp:224]: put it in dpct.cpp

File: clang/lib/Lex/PPDirectives.cpp [hunk 11]
```diff
@@ -717,7 +717,13 @@ void Preprocessor::SkipExcludedConditionalBlock(SourceLocation HashTokenLoc,
...
+          IsEvaluatingExpression = true;
```
Comment (zhiweij1) [inline clang/lib/Lex/PPDirectives.cpp]: move these 3 line in a function

File: clang/test/dpct/predefined_macro_replacement.cu [hunk 12]
```diff
@@ -173,3 +173,89 @@ int foo2(){
...
+//CHECK-NEXT: #elif !(SYCL_LANGUAGE_VERSION >= 202000) && AAA
+//CHECK-NEXT:   int2 a4;
+//CHECK-NEXT: #endif
+//CHECK-NEXT: #if BBB
+//CHECK-NEXT: #elif ((SYCL_LANGUAGE_VERSION >= 202000) ? 1 : 0) && AAA
+//CHECK-NEXT:   sycl::int2 a5;
+//CHECK-NEXT: #endif
+//CHECK-NEXT: }
+void foo4() {
+#define BBB 0
```
Comment (zhiweij1) [inline clang/test/dpct/predefined_macro_replacement.cu]: check when BBB is 1

Comment (zhiweij1) [reply clang/test/dpct/predefined_macro_replacement.cu]: done

File: clang/lib/DPCT/DPCT.cpp (hunk 10 above)
Comment (zhiweij1) [reply clang/lib/DPCT/DPCT.cpp:224]: CudaNameVersionMap is in Cuda.cpp, cannot move to this file

File: clang/lib/Lex/PPDirectives.cpp (hunk 11 above)
Comment (zhiweij1) [reply clang/lib/Lex/PPDirectives.cpp]: done

File: clang/lib/Lex/PPExpressions.cpp (hunk 9 above)
Comment (zhiweij1) [reply clang/lib/Lex/PPExpressions.cpp]: done

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**